### PR TITLE
Add decorator to mark middleware as async; fix middleware tests

### PR DESCRIPTION
**Pull request by timb07**

### CODE REVIEW

File: django/utils/decorators.py [hunk 1]
```diff
@@ -152,6 +152,18 @@ def callback(response):
     return _make_decorator
 
 
+def async_middleware(func):
+    """
+    Mark a middleware factory as returning an async middleware.
+    """
+    @wraps(func)
+    def wrapper(request):
+        return func(request)
```
Comment (andrewgodwin) [inline django/utils/decorators.py]: Do you need to do this? I think you can just have this be `func._is_async = True; return func`, no?

File: tests/middleware_exceptions/tests.py [hunk 2]
```diff
@@ -149,7 +149,7 @@ def test_log(self):
         with self.assertLogs('django.request', 'DEBUG') as cm:
             self.client.get('/middleware_exceptions/view/')
         self.assertEqual(
-            cm.records[0].getMessage(),
+            cm.records[1].getMessage(),
```
Comment (andrewgodwin) [inline tests/middleware_exceptions/tests.py:152]: Why did this index change? Just curious.

**Review by andrewgodwin — COMMENTED** [review]
Looks good overall, just a few niggles.

File: tests/middleware_exceptions/tests.py (hunk 2 above)
Comment (timb07) [reply tests/middleware_exceptions/tests.py:152]: The `Synchronous middleware adapted` log message is in `cm.records[0]`.

Comment (andrewgodwin) [reply tests/middleware_exceptions/tests.py:152]: Aha! Right.

**Review by andrewgodwin — APPROVED** [review]

**Comment by andrewgodwin** [comment]
Yup, this all looks good. Any more inbound, or should I merge?

**Comment by timb07** [comment]
> Yup, this all looks good. Any more inbound, or should I merge?

Nothing more inbound…

**Pull request merged**